File: Power Query/PQ_Challenge_141.R
library(tidyverse)
library(slider)
library(readxl)

input = read_excel("Power Query/PQ_Challenge_141.xlsx", range = "A1:C35")
test  = read_excel("Power Query/PQ_Challenge_141.xlsx", range = "E1:I35")

result = input %>%
  group_by(Month) %>%
  mutate(
    `3 Year MV` = slide_dbl(Defects, mean,  .after = -1, .before = 3, .complete = TRUE) %>% round(0),
    `5 Year MV` = slide_dbl(Defects, mean, .after = -1, .before = 5, .complete = TRUE) %>% round(0)
  ) %>% 
  ungroup()

identical(result, test)
#> [1] TRUE

File: Power Query/100-199/141/PQ_Challenge_141.py
import pandas as pd

input_data = pd.read_excel("PQ_Challenge_141.xlsx", usecols="A:C", nrows=35)
test = pd.read_excel("PQ_Challenge_141.xlsx", usecols="E:I", nrows=35)

result = input_data.copy()
result["3 Year MV"] = (
    result.groupby("Month")["Defects"]
    .transform(lambda s: s.rolling(window=4, min_periods=4).mean().round(0).shift(-1))
)
result["5 Year MV"] = (
    result.groupby("Month")["Defects"]
    .transform(lambda s: s.rolling(window=6, min_periods=6).mean().round(0).shift(-1))
)

print(result.equals(test))
